networkx/algorithms/coloring/greedy_coloring.py

```diff
@@ -46,29 +48,66 @@ def strategy_random_sequential(G, colors):
     return nodes
 
 
-# TODO There is a faster implementation of this algorithm: keep track of
-# the degrees and update them manually each time a node is removed.
 def strategy_smallest_last(G, colors):
     """Returns a list of the nodes of ``G``, "smallest" last.
 
-    Specifically, the node of minimum degree is repeatedly popped from
-    the graph, then this order is reversed.
+    Specifically, the degrees of each node are tracked. From this,
+    the node of minimum degree is repeatedly popped from the graph,
+    updating its neighbors' degrees, then this order is reversed. 
 
     ``G`` is a NetworkX graph. ``colors`` is ignored.
 
-    This implementation of the strategy runs in `O(n^2)` time.
+    This implementation of the strategy runs in `O(|V| + |E|)` time.
 
     This strategy is related to :func:`strategy_independent_set`: if we
     interpret each node removed as an independent set of size one, then
     this strategy chooses an independent set of size one instead of a
     maximal independent set.
 
+    A previous version of networkx noted that this faster algorithm
+    existed; this is the implementation.
+
+    ~~~~~~
+
+    Implementation details:
+    - Sets are used rather than lists for quick random-access deletes.
+    - In Python, pop() is deterministic for ints, which are the node IDs.
+      The behavior is reproducible in different invokations of Python. 
+
     """
     H = G.copy(with_data=False)
 
+    # Build initial degree list
+    degrees = defaultdict(set)
+    for (node, degree) in H.degree():
+        degrees[degree].add(node)
+
+    def move_to_lower_bucket(node):
+        degree = H.degree(node)
+        degrees[degree].remove(node)
+        # Clean up degree list.
+        if not degrees[degree]:
+            del degrees[degree]
+        degrees[degree - 1].add(node)
+
+    def pop_min_degree_node():
+        # Identify min-degree. (Needed for cleaning.)
+        min_degree = min(degrees)
+        # Retrieve node with min degree.
+        node = degrees[min_degree].pop()
+        # Clean up the degree list.
+        if not degrees[min_degree]:
+            del degrees[min_degree]
+        return node
+
+    def update_degree_list(removed_node):
+        for neighbor in H.neighbors(removed_node):
+            move_to_lower_bucket(neighbor)
+
     def min_and_pop():
```

Comment:
Fixed in 976ae17. I originally had that code (ignoring variable naming differences), but after seeing how pretty yours was, I decided to break it down.